packages/block-library/src/image/index.php

```diff
@@ -0,0 +1,357 @@
+ * The global $block_width array is expectd to be set by the theme for each block container.
+ * It should contain three values: default, wide and full, in pixels.
+ * - The `default` value should be the expected block width (similarly to $content_width).
+ * - The `wide` value is optional and is used when the block alignment is set to `wide`.
+ * - Similarly the `full` value is optional and used then the alignment is set to `full`.
+ * If `wide` and `full` are not set, the `default` value is used instead.
+ *
+ * Example:
+ *     $GLOBALS['block_width'] = array(
+ *         'default' => 640,
+ *         'wide'    => 800,
+ *         'full'    => 1024,
+ *     );
+ *
+ * In addition the $block_width array should be set contextually for each block container.
+ * For example: in the main content column the `default` width will be something like 640(px),
+ * but for a sidebar it would be something like 250.
+ *
+ * @since 4.2.0
+ *
+ * @return array Normalized array of expected block width with three elements: `default`, `wide`, and `full`.
+ */
+function gutenberg_get_block_width() {
+	global $content_width, $block_width;
+
+	// Get the width from the $block_width global array
+	// or fall back to using $content_width.
+	if ( empty( $block_width ) || ! is_array( $block_width ) || empty( $block_width['default'] ) ) {
+		if ( empty( $content_width ) ) {
+			$default_width = 1024; // Fall back to the "large" image file width.
+		} else {
+			$default_width = $content_width;
+		}
+
+		$width = array(
+			'default' => $default_width,
+			'wide' => $default_width,
+			'full' => $default_width,
+		);
+	} else {
+		$width = $block_width;
+
+		if ( empty( $width['wide'] ) ) {
+			$width['wide'] = $width['default'];
+		}
+
+		if ( empty( $width['full'] ) ) {
+			$width['full'] = $width['default'];
+		}
+	}
+
+	return $width;
+}
+
+/**
+ * Calculates the image width and height based on $block_witdh and the `scale` block attribute.
+ *
+ * @since 4.2.0
+ *
+ * @param array $block_attributes The block attributes.
+ * @param array $image_meta Optional. The image attachment meta data.
+ * @return array|bool An array of the image width and height, in that order, or false if the image data is missing from $block_attributes.
+ */
+function gutenberg_get_image_width_height( $block_attributes, $image_meta = null ) {
+	$block_witdh = gutenberg_get_block_width();
+
+	if ( empty( $block_attributes['fileWidth'] ) || empty( $block_attributes['fileHeight'] ) ) {
+		if ( ! empty( $block_attributes['width'] ) && ! empty( $block_attributes['height'] ) ) {
+			// Old post with an image that was resized?
+			return array(
+				$block_attributes['width'],
+				$block_attributes['height'],
+			);
+		} else {
+			// Image data is missing.
+			return false;
+		}
+	} else {
+		$image_file_width = (int) $block_attributes['fileWidth'];
+		$image_file_height = (int) $block_attributes['fileHeight'];
+	}
+
+	if ( ! empty( $image_meta ) &&
+		$image_file_width < $image_meta['width'] &&
```

Comment:
This seems overly complex. Am I missing cases other than the following:

1. See if this is the original size image
2. See if there is a large size image that's the same ratio
3. Use the large size dimensions instead of the original dimensions

If there are edge cases being accounted for beyond that, it would be good to add inline comments above.